File: euler112InManyLanguages/euler112.js

function isIncreasing(number){
    numAsStr = number.toString()
    var len = numAsStr.length
    if (len == 1){
        return true
    }
    if (numAsStr[len-1] >= numAsStr[len-2]){
        return isIncreasing(numAsStr.slice(0,-1))
    }
    return false
}

function isDecreasing(number){
    numAsStr = number.toString()
    var len = numAsStr.length
    if (len == 1){
        return true
    }
    if (numAsStr[len-1] <= numAsStr[len-2]){
        return isDecreasing(numAsStr.slice(0,-1))
    }
    return false
}

function isBouncy(number){
    return (!isDecreasing(number) && ! isIncreasing(number))
}

counter = 0
for (let i = 0; i < 1000; i++) {
    
    if (isBouncy(i)){
        counter ++
    }
}
console.log(counter, "should be 525")

percentage = 0
general = 0
bouncy = 0
toCheck = 1
goal = 0.99
while (percentage < goal){
    if (isBouncy(toCheck)){
        bouncy +=1
    }
    toCheck +=1
    general +=1
    percentage = bouncy / general
}
console.log("Answer to euler 112 is: ", general)

File: euler112InManyLanguages/euler112.java
package euler112InManyLanguages;

class euler112{

    public static boolean isIncreasing(int number){
        if (number < 10){
            return true;
        }
        while (number > 0){
            if (number < 10){
                return true;
            }
            if (number %10  < (number /10)%10){
                return false;
            }
            number = number / 10;
        }
        return true;
    }

    public static boolean isDecreasing(int number){
        if (number < 10){
            return true;
        }
        while (number > 0){
            if (number < 10){
                return true;
            }
            if (number %10  > (number /10)%10){
                return false;
            }
            number = number / 10;
        }
        return true;
    }

    private static boolean isBouncing(int number){
        return (!isDecreasing(number)&&!isIncreasing(number));
    }
    public static void main(String args[]){

        int counter = 0;
        for (int i = 0; i < 1000; i++) {
            if (isBouncing(i)){
                counter++;
            }
        }
        System.out.println(counter);

        double percentage = 0;
        double general = 0;
        double bouncy = 0;
        int toCheck = 1;
        double goal = 0.99;
        while (percentage < goal){
            if (isBouncing(toCheck)){
                bouncy +=1;
            }
            
            toCheck +=1;
            general +=1;
            percentage = bouncy / general;

        }
        System.out.println("Answer to euler 112 is: " +  general);

    }
}